File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int N = nums.length;
        int[] result = new int[N];
        
        int product = 1;
        for(int i =0; i < N; i++)
        {
            result[i] = product;
            product = product * nums[i];
        }
        
        product = 1;
        for(int i = N - 1; i >= 0; i--)
        {
            result[i] *= product;
            product = product * nums[i];
        }
        
        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = [1] * len(nums)

        for i in range(1, len(nums)):
            output[i] = output[i-1] * nums[i-1]

        suffix = 1
        for i in range(len(nums) - 1, -1, -1):
            output[i] *= suffix
            suffix *= nums[i]

        
        return output

        